### PR TITLE
drivers: can: make the fake CAN driver generally available

### PR DESCRIPTION
**Pull request by henrikbrixandersen**
Make the fake CAN controller driver available for use in tests outside of the CAN shell test and start a list of FFF-based fake drivers available in Zephyr in the documentation.

### CODE REVIEW

**Comment by henrikbrixandersen** [comment]
What do people think about providing fake drivers in the regular `drivers/` tree for use as stubs or mocks in both in-tree and out-of-tree tests?

**Comment by cfriedt** [comment]
> What do people think about providing fake drivers in the regular `drivers/` tree for use as stubs or mocks in both in-tree and out-of-tree tests?

@henrikbrixandersen  Great idea. Additionally, it's possible to have emulated peripherals on the fake CAN bus as well, if that's a thing (I'm less accustomed to CAN so maybe not). I would suggest calling them `<periph>_emul` if it's possible and they can be inserted into device tree like any other device instance.

**Review by gmarull — COMMENTED** [review]
I think it is a good idea to provide mocks/fakes for existing driver interfaces. This will allow writing unit tests more easily for the API clients.

**Comment by henrikbrixandersen** [comment]
> Also, did the first two commits land here by accident?

They did, thanks.

**Comment by henrikbrixandersen** [comment]
> @henrikbrixandersen - is it potentially better to use the `*_emul` convention here rather than `*_fake`? I originally used `*_fake` for `gpio_emul` even though fake might have been a better description. It's really just for consistence.

The proposal is for fake drivers, not emulators nor simulators. Fakes as in FFF-based empty shells that can be used either for simple stubs or mocks in unit tests.

**Comment by henrikbrixandersen** [comment]
> @henrikbrixandersen Great idea. Additionally, it's possible to have emulated peripherals on the fake CAN bus as well, if that's a thing (I'm less accustomed to CAN so maybe not). I would suggest calling them `<periph>_emul` if it's possible and they can be inserted into device tree like any other device instance.

Again, this is not an emulator (and CAN is more like a network than a peripheral bus).

**Comment by mbolivar-nordic** [comment]
Architecture WG:

Concept is OK, some requests from @nashif for potential documentation changes.

**Comment by henrikbrixandersen** [comment]
This is ready for review.

**Review by mbolivar-nordic — DISMISSED** [review]

**Review by martinjaeger — DISMISSED** [review]

**Review by gmarull — DISMISSED** [review]

**Review by str4t0m — DISMISSED** [review]

**Comment by henrikbrixandersen** [comment]
Rebased to fix merge conflict in `drivers/can/Kconfig` after 58e12726bfe85146a4a75b21372c17db0180039f.

**Review by gmarull — DISMISSED** [review]

File: doc/develop/test/ztest.rst [hunk 1]
```diff
@@ -489,6 +489,11 @@ following in your source::
 
     #include <zephyr/fff.h>
 
+Zephyr provides several FFF-based fake drivers which can be used as either stubs or mocks. Fake
+driver instances are configured via :ref:`devicetree` and :ref:`kconfig`. See the following
+devicetree bindings for more information:
+
+ - :dtcompatible:`zephyr,fake-can`
```
Comment (gmarull) [inline doc/develop/test/ztest.rst:496]: To avoid bitrot: I'd suggest saying here something like
```rst
look for ``zephyr,fake-*``
```

Comment (henrikbrixandersen) [reply doc/develop/test/ztest.rst:496]: That's nice too, but having a direct reference makes it easier to find the relevant compatibles.

I am pondering creating a separate documentation page under testing, which describes the various fake drivers, emulated drivers, and simulators available in Zephyr - their design patterns and suggested use-cases. I'd like to handle that outside this PR, though.

**Review by str4t0m — DISMISSED** [review]

**Comment by henrikbrixandersen** [comment]
CI issue unrelated, addressed here: https://github.com/zephyrproject-rtos/zephyr/pull/51888

**Comment by henrikbrixandersen** [comment]
Rebased to pull in b3b64e9edef256bb1ed837d6947761ec6b851059 in order to fix CI.

**Review by martinjaeger — APPROVED** [review]

**Review by str4t0m — APPROVED** [review]

**Pull request merged**